Generate smithy-build.json from sbt config

daddykotex opened the pull request:
This is not perfect, but I guess it's a good first step and it might help our user setup their original configuration.

This current version has no support for customization though. Should I do that in this PR or we're okay revisiting this when required.

for sbt: `sbt generateSmithyBuildJson`

for mill: `mill  smithy4s.codegen.mill.Smithy4sExternalModule/generateSmithyBuildJson`

Fixes #497

daddykotex committed:
Dummy implementation to generate smithy-build.json file

daddykotex committed:
Update smithy-build.json

daddykotex committed:
Write smithy-build.json as json

daddykotex committed:
Remove duplicated import

Baccata commented on modules/codegen-plugin/src/smithy4s/codegen/Smithy4sCodegenPlugin.scala:
```diff
@@ -255,6 +263,24 @@ object Smithy4sCodegenPlugin extends AutoPlugin {
+      val repositories = externalResolvers.value.collect(prepareResolvers)
+      SmithyBuildJson.toJson(imports, deps, repositories)
+    },
+    config / smithy4sGeneratedSmithyBuildFile := {
+      val out = baseDirectory.value / "smithy-build.json"
+      IO.write(out, (config / smithy4sGeneratedSmithyBuild).value)
+      out
```
not sure this cuts it : ideally we want to aggregate all dependencies from all modules and create the file from this aggregation.

kubukoz replied:
wish the LSP supported multiple config files in a workspace :/

contributions welcome I guess

daddykotex commented:
For now, I'll investigate on the following alternatives to improve the current implementation:

* the task is global to all the modules and will merge the results from all modules where the smithy4s codegen plugins is enabled
* the task in each module that has the plugin enabled will read the current smithy-build file and append to it if necessary (aware of concurrency)

daddykotex committed:
Merge remote-tracking branch 'origin/series/0.18' into dfrancoeur/497-smithy-build

daddykotex commented on modules/codegen-plugin/src/sbt-test/codegen-plugin/defaults/build.sbt:
```diff
@@ -5,5 +5,41 @@ lazy val root = (project in file("."))
+                        |    ],
+                        |    "maven": {
+                        |        "dependencies": [
+                        |            "com.disneystreaming.alloy:alloy-core:0.2.6"
```
just realized that this test will be sensitive to dependency upgrade (right now alloy is bumped in 0.18 and after merged series/0.18 here, it fails :/ )

kubukoz replied:
TBH I wouldn't check the file against a text version of JSON in the first place. I wouldn't want this to break because e.g. jsoniter changes ordering of serialized fields, or if we change config.

Just checking whether the file contains relevant strings should do (e.g. the `com.disneystreaming.alloy:alloy-core` prefix, `src/main/smithy`, `1.0`), and maybe also verifying that it all parses as JSON.

If you want to be more strict I'd still compare JSON ASTs rather than strings, and then the version numbers could maybe be ignored

kubukoz commented on modules/codegen-plugin/src/smithy4s/codegen/Smithy4sCodegenPlugin.scala:
```diff
@@ -290,6 +316,22 @@ object Smithy4sCodegenPlugin extends AutoPlugin {
       )
     )
 
+  private def prepareResolvers: PartialFunction[Resolver, String] = {
```
nitpick: `val`

kubukoz commented on modules/codegen-plugin/src/smithy4s/codegen/Smithy4sCodegenPlugin.scala:
```diff
@@ -255,6 +263,24 @@ object Smithy4sCodegenPlugin extends AutoPlugin {
+        .filter(_.configurations.exists(_.contains(Smithy4s.name)))
+        .flatMap(moduleIdEncode(_, scalaBin))
+      val imports = (config / smithy4sInputDirs).value
+        .collect(prepareInputDirs(baseDir, sourceManagedDir))
```
suggestion: named parameters for `baseDir` / `sourceManagedDir`

kubukoz commented on modules/codegen-plugin/src/smithy4s/codegen/Smithy4sCodegenPlugin.scala:
```diff
@@ -290,6 +316,22 @@ object Smithy4sCodegenPlugin extends AutoPlugin {
+      base: File,
+      sourceManaged: File
+  ): PartialFunction[File, String] = {
+    // exclude files that are under sourceManaged
```
question: can you explain why we wanna do this? I was thinking that if these files are in `smithy4sInputDirs` and affect codegen, we'd want to import them

daddykotex replied:
it's a good question. when I saw it in the list, the first time I generated it
my brain automatically marked it as unwanted because it was under `target`

maybe the fact that they are generated and user modifications are discarded because the build tool handles those files

I can keep it no problem though

daddykotex replied:
kept it

kubukoz replied:
what files are those? the ones that set metadata for the Scala import syntax, perchance?

daddykotex replied:
```
> tree bar/target/scala-2.13/src_managed/main/smithy/
bar/target/scala-2.13/src_managed/main/smithy/
└── generated-metadata.smithy
> cat bar/target/scala-2.13/src_managed/main/smithy/generated-metadata.smithy 
$version: "2"
metadata smithy4sWildcardArgument = "_"
metadata smithy4sRenderOptics = false
```

daddykotex committed:
Expose `smithy4sGeneratedSmithyBuildMerge` which read/write to existing file

This method is allows you to generate the smithy-build.json file for each
module and then merge the results together by reading/writing the result to
the file for each module

daddykotex committed:
Avoid two tasks, one is fine and fix tests

daddykotex committed:
Simplify implementation

The `task / aggregate := false` does not seem to do what it says
it would.

by taking the content Option[String] out of the makeSmithyBuild,
I can avoid the taskDyn

daddykotex committed:
Add test and fix broken one

daddykotex commented on modules/codegen-plugin/src/smithy4s/codegen/Smithy4sCodegenPlugin.scala:
```diff
@@ -290,6 +314,59 @@ object Smithy4sCodegenPlugin extends AutoPlugin {
+
+    val smithyBuildJson = makeSmithyBuild(config).value
+
+    synchronized {
```
it seems like `task / aggregate := false` did not prevent parallel execution as I had hoped and sometimes I was left with a broken json file. so I slapped a synchronized

kubukoz replied:
tbh that sounds like a bit of a band aid rather than a long-term solution. I'd rather have one top-level task that you run on your `root` project, which eliminates any concurrency in side-effecting IO

daddykotex replied:
we discussed this internally, and I don't know how to implement such a task

especially not in a sbt plugin

maybe @keynmol would know?

context: I'd like a task that I can run on the root module (or any aggregate module) that runs a task defined in a plugin. ideally, the task would aggregate the result of all the task executions, and then write to a file. the idea is that every module that has `Smithy4sCodegenPlugin` has a task to generate a `smithy-build.json` definition, but we'd like to have one task to reduce them all together and write them to disk

daddykotex replied:
got rid of that 👍

daddykotex committed:
Add info to requirement failed message

daddykotex committed:
Version on CI is different

daddykotex committed:
Generate smithy-build.json for mill builds

This generate one file, for all modules:

```json
{
    "version": "1.0",
    "imports": [
        "bar/src/main/smithy",
        "foo/src/main/smithy"
    ],
    "maven": {
        "dependencies": [
            "com.disneystreaming.alloy:alloy-core:0.2.7"
        ],
        "repositories": []
    }
}
```

daddykotex committed:
Implement sbt with a command, rename mill command

daddykotex committed:
Merge remote-tracking branch 'origin/series/0.18' into dfrancoeur/497-smithy-build

daddykotex committed:
Remove leftover println

Baccata commented on modules/mill-codegen-plugin/src/smithy4s/codegen/mill/Smithy4sExternalModule.scala:
```diff
@@ -0,0 +1,101 @@
+import mill.define.Command
+import mill.api.PathRef
+
+object Smithy4sExternalModule extends ExternalModule {
```
when creating mill external module, it's important to think about the UX imposed onto the users, as they don't get autocompletion. You want the command to be as easy to remember as possible. I think it should be something like : 

`smithy4s.LSP/config`

daddykotex replied:
That would mean moving the object under the `smithy4s` package, right?

daddykotex replied:
would `smithy4s.codegen.LSP/smithyBuild` be fine with you?

I feel like we're gonna add other commands in the future in this module and `config` sounds too generic

I added `smithy4s.codegen` so package visiblity of the `SmithyBuildJson` still make some sense

Baccata replied:
package is fine. For reference, similar commands in the mill ecosystem are : 

``` 
mill.idea.GenIdea/idea 
mill.contrib.bloop.Bloop/install
```

I think the first one is silly, there's 3 repetitions of `idea`. I however like the `GenIdea` object
I like `install` but it's a bit generic. 

Combining the two, we could have something like : 

``` 
smithy4s.codegen.LSPConfig/update
smithy4s.codegen.LSP/updateConfig
```

NB : after a decision on this, the SBT command should be updated to be similar. `smithy4sUpdateLSPConfig` would be alright.

daddykotex replied:
done, I've added some docs in the editor support page as well

Baccata commented on modules/mill-codegen-plugin/src/smithy4s/codegen/mill/Smithy4sModule.scala:
```diff
@@ -38,23 +38,23 @@ import mill.scalalib.CrossVersion.Full
 trait Smithy4sModule extends ScalaModule {
 
   /** Input directory for .smithy files */
-  protected def smithy4sInputDirs: Target[Seq[PathRef]] = T.sources {
```
just make them package private

Baccata commented on modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala:
```diff
@@ -0,0 +1,81 @@
+ *  limitations under the License.
+ */
+
+package smithy4s.codegen.internals
```
all the constructs in here should probably be package-private (`private [smithy4s]`)

daddykotex replied:
yeah, all of that should be private, because I hide all of that functionality behind the `SmithyBuildJson` object

Baccata committed:
Rewrite mill LSP config task more idiomatically, add unit-tests

Baccata commented:
I've taken the liberty of rewriting the mill integration more idiomatically, and added some tests : [7d60f5c](https://github.com/disneystreaming/smithy4s/pull/1208/commits/7d60f5ccf9d72d4d85f9add2a965e1164628a1bf)

daddykotex commented on modules/mill-codegen-plugin/src/smithy4s/codegen/mill/Smithy4sExternalModule.scala:
```diff
@@ -21,81 +21,48 @@ import mill.define.ExternalModule
-          .map(_.toString)
-          .distinct
-      }
+    val depsTask = Target
```
I had a similar approach at first and I rewrote to avoid traversing the s4smodule seq multiple time (like I did in sbt)

Baccata replied:
Yeah, the approach you had was similar to chaining IO using `unsafeRunSync`, which ain't great. 

I can't comment much on whether the SBT approach is idiomatic because I don't know SBT as well 😅

daddykotex commented on modules/mill-codegen-plugin/test/src/smithy4s/codegen/mill/SmithyLSPConfigSpec.scala:
```diff
@@ -0,0 +1,58 @@
+
+  private object testKit extends MillTestKit
+
+  test("config gets generated") {
```
I did not write the test because I was not sure how to, that `root` trick does it perfectly

thanks for stepping up

daddykotex commented on modules/mill-codegen-plugin/test/src/smithy4s/codegen/mill/SmithyLSPConfigSpec.scala:
```diff
@@ -0,0 +1,58 @@
+          |  "imports": ["./foo/smithy"],
+          |  "maven": {
+          |    "dependencies": [
+          |       "com.disneystreaming.alloy:alloy-core:${smithy4s.codegen.BuildInfo.alloyVersion}",
```
clever, the version is in the `BuildInfo`

daddykotex committed:
Adjust visiblity on SmithyBuild* objects

kubukoz commented on modules/mill-codegen-plugin/test/src/smithy4s/codegen/mill/SmithyLSPConfigSpec.scala:
```diff
@@ -0,0 +1,58 @@
+import mill.testkit.MillTestKit
+import sourcecode.FullName
+
+class SmithyLSPConfigSpec extends munit.FunSuite {
```
thought: to be precise it's the smithy-cli config, the LSP happens to use it

Baccata replied:
<a href="https://imgflip.com/i/80to1e"><img src="https://i.imgflip.com/80to1e.jpg" title="made at imgflip.com"/></a><div></div>

:trollface:

kubukoz replied:
I missed this 😂 

it was a good comment. If you use smithy-cli (which has some nice things in it), you won't think to use `smithyLspConfig`

kubukoz reviewed (commented):
I like the idea to merge with the existing files, that should allow custom options, right?

daddykotex commented:
> I like the idea to merge with the existing files, that should allow custom options, right?

absolutely, it does

there is a sbt-test that tests that

daddykotex committed:
Merge remote-tracking branch 'origin/series/0.18' into dfrancoeur/497-smithy-build

daddykotex committed:
Rename commands for sbt/mill to generate lsp config
